### PR TITLE
Change reporters to use a specified output and error stream. Add tests for output.

### PR DESCRIPTION
**Pull request by EricWF**
This patch attempts to make it easier to test the output of the reporters. Bugs in the output seems to pop up frequently and it would be nice to write unit tests for it.

This patch solves the test-ability problem by making all output use user-provided streams. Once this patch lands we can slowly add tests for the output under all the various benchmark options.

This branch is cloned from the 'SkipWithError(...)' branch so please ignore those changes here until it langs.

### CODE REVIEW

File: include/benchmark/reporter.h [hunk 1]
```diff
@@ -105,12 +110,45 @@ class BenchmarkReporter {
   // reported.
   virtual void Finalize();
 
+  // REQUIRES: 'out' shall be valid for the lifetime of the reporter.
+  // Set the stream used for non-error output to the value specified by 'out'.
+  void SetOutputStream(std::ostream& out) {
```
Comment (dominichamon) [inline include/benchmark/reporter.h]: google style: no non-const references. please take a pointer directly (and CHECK for nullptr. sorry).

File: include/benchmark/reporter.h [hunk 2]
```diff
@@ -105,12 +110,45 @@ class BenchmarkReporter {
...
+    output_stream_ = &out;
+  }
+
+  // REQUIRES: 'out' shall be valid for the lifetime of the reporter.
+  // Set the stream used for error output to the value specified by 'out'.
+  void SetErrorStream(std::ostream& out) {
```
Comment (dominichamon) [inline include/benchmark/reporter.h]: rename `out` to `err`?

File: include/benchmark/reporter.h [hunk 3]
```diff
@@ -105,12 +110,45 @@ class BenchmarkReporter {
...
+    error_stream_ = &out;
+  }
+
+  // REQUIRES: 'out' shall be valid for the lifetime of the reporter.
+  // Set the stream used for both error and non-error output to the value
+  // specified by 'out'.
+  void SetStreams(std::ostream& out) {
```
Comment (dominichamon) [inline include/benchmark/reporter.h]: not sure this is useful given you have the above...

File: src/console_reporter.cc [hunk 4]
```diff
@@ -32,26 +34,36 @@ namespace benchmark {
 bool ConsoleReporter::ReportContext(const Context& context) {
   name_field_width_ = context.name_field_width;
 
-  std::cerr << "Run on (" << context.num_cpus << " X " << context.mhz_per_cpu
+  auto& Out = GetOutputStream();
+  auto& Err = GetErrorStream();
+
+#ifdef BENCHMARK_OS_WINDOWS
+  if (FLAGS_color_print && &Out != &std::cout) {
```
Comment (dominichamon) [inline src/console_reporter.cc]: what if Out is std::cerr?

File: include/benchmark/reporter.h (hunk 1 above)
Comment (EricWF) [inline include/benchmark/reporter.h]: Ack. Does this apply to the getters as well?

File: include/benchmark/reporter.h (hunk 3 above)
Comment (EricWF) [inline include/benchmark/reporter.h]: Yeah. I'll remove this.

File: include/benchmark/reporter.h (hunk 1 above)
Comment (dominichamon) [inline include/benchmark/reporter.h]: doesn't seem to, just arguments.

ftr, i don't like this rule, but there it is.

File: src/colorprint.cc [hunk 5]
```diff
@@ -74,14 +78,53 @@ PlatformColorCode GetPlatformColorCode(LogColor color) {
...
+
 }  // end namespace
 
-void ColorPrintf(LogColor color, const char* fmt, ...) {
+std::string FormatString(const char *msg, va_list args) {
+  // we might need a second shot at this, so pre-emptivly make a copy
+  va_list args_cp;
+  va_copy(args_cp, args);
+
+  //TODO, use std::array for first attempt to avoid one memory allocation
```
Comment (EricWF) [inline src/colorprint.cc]: Note to self: remove this comment since it's been implemented.

File: src/console_reporter.cc (hunk 4 above)
Comment (EricWF) [inline src/console_reporter.cc]: It could likely be made to work but I have no way of testing on Windows so I avoided changing it.

**Comment by dominichamon** [comment]
can you resolve conflicts so i can merge please?

**Pull request merged**